src/ContractionTreeToGraph.jl

```diff
@@ -0,0 +1,120 @@
+        spawn_child_branch(contract_sequence[2], current_g, leaves)
+    else
+        #If it is it is just a vertex
+        add_vertex!(current_g, ([contract_sequence], setdiff(leaves, [contract_sequence])))
+    end
+end
+
+"""Utility functions for the graphical representation of a contraction sequence
+Perhaps it should be a specific object or type of tree?!"""
+
+"""Determine if a node is a leaf"""
+function leaf_node(g::AbstractGraph, v)
+    return length(neighbors(g, v)) == 1
+end
+
+"""Determine if an edge involves a leaf (at src or dst)"""
+function leaf_edge(g::AbstractGraph, e)
+    return leaf_node(g, src(e)) || leaf_node(g, dst(e))
+end 
+
+"""Determine if a node has no neighbours which are leaves"""
+function no_leaf_neighbours(g::AbstractGraph, v)
+    for vn in neighbors(g, v)
+        if(leaf_node(g, vn))
+            return false
+        end
+    end
+    return true
+end
+
+"""Get all edges which do not involve a leaf
+(why does edges(g)[contraction_edge(g, edges(g)) .== true] not work?)"""
+function non_leaf_edges(g::AbstractGraph)
+    return edges(g)[findall(==(1), [!leaf_edge(g,e) for e in edges(g)])]
+end
+
+"""Get all nodes which aren't leaves and don't have leaf neighbours (these represent internal points in the contraction sequence)"""
+function internal_nodes(g::AbstractGraph)
+    return vertices(g)[findall(==(1), [no_leaf_neighbours(g, v) && !leaf_node(g, v) for v in vertices(g)])]
+end
+
+"""Get all nodes which have leaf neighbours but aren't leaves themselves (these represent start/ end points of a contraction sequence)"""
+function external_nodes(g::AbstractGraph)
+    return vertices(g)[findall(==(1), [!no_leaf_neighbours(g, v) && !leaf_node(g,v)  for v in vertices(g)])]
+end
+
+"""Get the vertex bi-partition that a given edge between non-leaf nodes represents"""
+function edge_bipartition(g::AbstractGraph, e)
+
+    if(leaf_edge(g, e))
+        println("ERROR: EITHER THE SOURCE OR THE VERTEX IS A LEAF SO EDGE DOESN'T REALLY REPRESENT A BI-PARTITION")
+    end
+
+    vsrc_set, vdst_set = [Set(vni) for vni in src(e)], [Set(vni) for vni in dst(e)]
+    c1, c2, c3 = [src(e)[1]..., src(e)[2]...], [src(e)[2]..., src(e)[3]...], [src(e)[1]..., src(e)[3]...]
+    left_bipartition = Set(c1) ∈ vdst_set ? c1 : Set(c2) ∈ vdst_set ? c2 : c3
+
+    c1, c2, c3 = [dst(e)[1]..., dst(e)[2]...], [dst(e)[2]..., dst(e)[3]...], [dst(e)[1]..., dst(e)[3]...]
+    right_bipartition = Set(c1) ∈ vsrc_set ? c1 : Set(c2) ∈ vsrc_set ? c2 : c3
+
+    return left_bipartition, right_bipartition
+end
+
+"""Given a contraction node, get the keys living on all its neighbouring leaves"""
+function external_node_keys(g::AbstractGraph, v)
+    return [Base.Iterators.flatten(v[findall(==(1), [length(vi) == 1 for vi in v])])...]
+end
+
+"""Given a contraction node, get all keys which are not living on a neighbouring leaf"""
+function external_contraction_node_ext_keys(g::AbstractGraph, v)
```

Comment:
This is a bit confusing, it would help to get a more extended explanation for what this is used for.

Reply:
This is the opposite of the `external_node_keys` function (get all keys which aren't associated to neighbouring leaves).